File: Python/ShellSort.py
# Python3 program for implementation of Shell Sort
from typing import List


def shell_sort(arr: List[int]) -> None:
    # Start with a big gap, then reduce the gap
    n: int = len(arr)
    gap: int = n // 2

    # Do a gapped insertion sort for this gap size.
    # The first gap elements a[0..gap-1] are already in gapped
    # order keep adding one more element until the entire array
    # is gap sorted
    while gap > 0:
        for i in range(gap, n):
            # add a[i] to the elements that have been gap sorted
            # save a[i] in temp and make a hole at position i
            temp: int = arr[i]

            # shift earlier gap-sorted elements up until the correct
            # location for a[i] is found
            j: int = i
            while j >= gap and arr[j - gap] > temp:
                arr[j] = arr[j - gap]
                j -= gap

            # put temp (the original a[i]) in its correct location
            arr[j] = temp
        gap //= 2


# Driver code to test above
if __name__ == '__main__':
    arr = [12, 34, 54, 2, 3]
    print("Array before sorting:", arr)

    shell_sort(arr)
    print("Array after sorting:", arr)

File: Cpp/ShellSort.cpp
#include<iostream>
using namespace std;
void swapping(int &a, int &b) {        
//to swap the content of a and b
   int temp;
   temp = a;
   a = b;
   b = temp;
}
void display(int *array, int size) {
//to display the elements in array
   for(int i = 0; i<size; i++)
      cout << array[i] << " ";
   cout << endl;
}
void shellSort(int *arr, int n) {
   int gap, j, k;
   for(gap = n/2; gap > 0; gap = gap / 2) { 
      decreasing by gap /2
      for(j = gap; j<n; j++) {
         for(k = j-gap; k>=0; k -= gap) {
            if(arr[k+gap] >= arr[k])
               break;
            else
               swapping(arr[k+gap], arr[k]);
         }
      }
   }
}
int main() {
   int n;
   cout << "Enter number of elements: ";
   cin >> n;
   //create an array with given number of elements
   int arr[n];     
   cout << "Enter elements:" << endl;
   for(int i = 0; i<n; i++) {
      cin >> arr[i];
   }
   cout << "Before Sorting: ";
   display(arr, n);
   shellSort(arr, n);
   cout << "After Sorting: ";
   display(arr, n);
}
